pkg/client/injection/reconciler/pipeline/v1beta1/pipelinerun/controller.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2020 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Code generated by injection-gen. DO NOT EDIT.
+
+package pipelinerun
```

Comment:
Our `controller.go` is not in an API specific folder, but it uses this genreconciler which is API specific.
We only store one version of the API, so I suppose this simply means that in the controller will use the reconciler for the API version that is stored?

Reply:
So controllers use the version of the "primary resource" that they set up informers for, which is orthogonal from the stored version.  This is cheapest when it matches the storage version because the informer will go through the conversion webhooks.

We used to (briefly) version our controller directories, but moved away from this when we realized that we'd only have a controller for the single canonical version and never multiple versions of the controller for a particular resource.

This is versioned because you could generate it for each version of a resource, and have multiple controllers[*] of a resource that consumer different versions of the generated code (likely during a version transition).

[*] - We do this in a few cases: extensibility (class-based reconciliation for plug-points), background processes that have been split off the main reconciliation loop (only one writes status).

Hopefully that clarifies, but if not I'd be happy to try to explain further.